clash-ghc/src-ghc/Clash/GHC/LoadModules.hs

```diff
@@ -454,58 +463,77 @@ loadModules useColor hdl modName dflagsM idirs = do
-              No top-level function called '#{topEntityName}' found. Did you
-              forget to export it?
-            |]
-          else
-            Panic.pgmError [I.i|
-              No top-level function called 'topEntity' found, nor a function with
-              a 'Synthesize' annotation in module #{modName1}. Did you forget to
-              export them?
-
-              For more information on 'Synthesize' annotations, check out the
-              documentation of "Clash.Annotations.TopEntity".
-            |]
-        ([], _) ->
-          return allSyn'
-        ([x], _) ->
-          case lookup x topSyn of
-            Nothing ->
-              case lookup x benchAnn of
-                Nothing -> return ((x,Nothing,listToMaybe benches):allSyn')
-                Just y  -> return ((x,Nothing,Just y):allSyn')
-            Just _ ->
-              return allSyn'
-        (_, _) ->
-          Panic.pgmError $ $(curLoc) ++ "Multiple 'topEntities' found."
+    let
+      -- All binders synthesized with Synthesize, all binders annotated with
+      -- TestBench and the binders they're pointing to, plus magically named
+      -- functions called "topEntity" or "testBench". Synthesized in case user
+      -- didn't specify a particular target.
+      allImplicit = nubSort $
+           Map.keys benchAnn
+        <> Map.keys allSyn
+        <> concat (Map.elems benchAnn)
+        <> filter ((=="topEntity") . varNameString) rootIds
+        <> filter ((=="testBench") . varNameString) rootIds
+        <> topSyn
+
+      -- Top entities we wish to synthesize. Users can filter these with -main-is.
+      topEntities1 =
+        case GHC.mainFunIs =<< dflagsM of
+          Just mainIsNm ->
+            -- Use requested top entity.
+            --
+            -- TODO: Look up associated test benches in 'benchAnn'. This would
+            --       be wasted effort if implemented right now, as 'getMainTopEntity'
+            --       would later remove them again. Functionality of that function
+            --       should be moved here.
+            --
+            -- TODO: Handle fully qualified names to -main-is
+            case find ((==mainIsNm) . varNameString) rootIds of
+              Nothing ->
+                Panic.pgmError [I.i|
+                  No top-level function called '#{mainIsNm}' found. Did you
+                  forget to export it?
+                |]
+              Just top ->
+                -- Note that we return /all/ top entities here, even the ones
+                -- we don't which to synthesize. 'getMainTopEntity' will later
+                -- restrict this to just this top entity (and its dependencies,
+                -- which is why we return everything in the first place).
+                --
+                -- This is quite wasteful though; als Clash will load all
+                -- definitions even though it will end up using just a few. TODO
+                nubSort (top:allImplicit)
```

Comment:
Why do we sort them lexicographically?

Reply:
The order they come in doesn't really matter as they get topsorted later. This is just to make sure that `top` doesn't appear twice in `top:allImplicit`, which might happen if the binder selected by `-main-is` has a Synthesize annotation. I don't use `nub` as it is `O(n2)`.